File: app/src/lib/validation/passwordPolicy.ts
/**
 * Centralized password complexity validation per FR-009c.
 * Aligned with backend PasswordPolicy.cs for consistent enforcement.
 */

export const PASSWORD_MIN_LENGTH = 8
export const PASSWORD_MAX_LENGTH = 128

export interface PasswordRequirement {
  id: string
  label: string
  met: boolean
}

/**
 * Evaluates password against all complexity requirements.
 * Returns array of requirement objects with met status for UI display.
 */
export function getPasswordRequirements(password: string): PasswordRequirement[] {
  return [
    {
      id: 'length',
      label: `At least ${PASSWORD_MIN_LENGTH} characters`,
      met: password.length >= PASSWORD_MIN_LENGTH,
    },
    {
      id: 'lowercase',
      label: 'Lowercase letter',
      met: /[a-z]/.test(password),
    },
    {
      id: 'uppercase',
      label: 'Uppercase letter',
      met: /[A-Z]/.test(password),
    },
    {
      id: 'number',
      label: 'Number',
      met: /\d/.test(password),
    },
    {
      id: 'special',
      label: 'Special character',
      met: /[^A-Za-z0-9]/.test(password),
    },
  ]
}

/**
 * Validates password against all complexity requirements.
 * @returns true if all requirements are met
 */
export function isPasswordValid(password: string): boolean {
  if (!password || password.length < PASSWORD_MIN_LENGTH) {
    return false
  }

  if (password.length > PASSWORD_MAX_LENGTH) {
    return false
  }

  return getPasswordRequirements(password).every((req) => req.met)
}

/**
 * Returns list of unmet requirement labels for error messaging.
 * Aligned with backend GetMissingRequirements for consistent error reporting.
 */
export function getMissingRequirements(password: string): string[] {
  const missing: string[] = []

  if (!password) {
    return [
      `Password must be at least ${PASSWORD_MIN_LENGTH} characters`,
      'Password must contain a lowercase letter',
      'Password must contain an uppercase letter',
      'Password must contain a number',
      'Password must contain a special character',
    ]
  }

  if (password.length < PASSWORD_MIN_LENGTH) {
    missing.push(`Password must be at least ${PASSWORD_MIN_LENGTH} characters`)
  }

  if (password.length > PASSWORD_MAX_LENGTH) {
    missing.push(`Password must not exceed ${PASSWORD_MAX_LENGTH} characters`)
  }

  if (!/[a-z]/.test(password)) {
    missing.push('Password must contain a lowercase letter')
  }

  if (!/[A-Z]/.test(password)) {
    missing.push('Password must contain an uppercase letter')
  }

  if (!/\d/.test(password)) {
    missing.push('Password must contain a number')
  }

  if (!/[^A-Za-z0-9]/.test(password)) {
    missing.push('Password must contain a special character')
  }

  return missing
}

File: Server/ClinicalIntelligence.Api/Services/Auth/PasswordPolicy.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace ClinicalIntelligence.Api.Services.Auth;

/// <summary>
/// Centralized password complexity validation per FR-009c.
/// Provides reusable validation logic with detailed requirement feedback.
/// </summary>
public static class PasswordPolicy
{
    public const int MinLength = 8;
    public const int MaxLength = 128;

    private static readonly Regex LowercaseRegex = new(@"[a-z]", RegexOptions.Compiled);
    private static readonly Regex UppercaseRegex = new(@"[A-Z]", RegexOptions.Compiled);
    private static readonly Regex DigitRegex = new(@"\d", RegexOptions.Compiled);
    private static readonly Regex SpecialCharRegex = new(@"[^A-Za-z0-9]", RegexOptions.Compiled);

    /// <summary>
    /// Validates password against all complexity requirements.
    /// </summary>
    /// <param name="password">The password to validate.</param>
    /// <returns>True if password meets all requirements; otherwise false.</returns>
    public static bool IsValid(string? password)
    {
        return GetMissingRequirements(password).Count == 0;
    }

    /// <summary>
    /// Returns a list of unmet password requirements for error reporting.
    /// </summary>
    /// <param name="password">The password to validate.</param>
    /// <returns>List of requirement descriptions that are not met.</returns>
    public static List<string> GetMissingRequirements(string? password)
    {
        var missing = new List<string>();

        if (string.IsNullOrEmpty(password))
        {
            missing.Add($"Password must be at least {MinLength} characters");
            missing.Add("Password must contain a lowercase letter");
            missing.Add("Password must contain an uppercase letter");
            missing.Add("Password must contain a number");
            missing.Add("Password must contain a special character");
            return missing;
        }

        if (password.Length < MinLength)
        {
            missing.Add($"Password must be at least {MinLength} characters");
        }

        if (password.Length > MaxLength)
        {
            missing.Add($"Password must not exceed {MaxLength} characters");
        }

        if (!LowercaseRegex.IsMatch(password))
        {
            missing.Add("Password must contain a lowercase letter");
        }

        if (!UppercaseRegex.IsMatch(password))
        {
            missing.Add("Password must contain an uppercase letter");
        }

        if (!DigitRegex.IsMatch(password))
        {
            missing.Add("Password must contain a number");
        }

        if (!SpecialCharRegex.IsMatch(password))
        {
            missing.Add("Password must contain a special character");
        }

        return missing;
    }

    /// <summary>
    /// Validates password and throws InvalidOperationException if invalid.
    /// Error message does not include the password value for security.
    /// </summary>
    /// <param name="password">The password to validate.</param>
    /// <param name="contextMessage">Optional context for the error message (e.g., "ADMIN_PASSWORD").</param>
    /// <exception cref="InvalidOperationException">Thrown when password does not meet requirements.</exception>
    public static void ValidateOrThrow(string? password, string contextMessage = "Password")
    {
        var missing = GetMissingRequirements(password);
        if (missing.Count > 0)
        {
            throw new InvalidOperationException(
                $"{contextMessage} must be at least {MinLength} characters with mixed case, number, and special character");
        }
    }

    private const string LowercaseChars = "abcdefghijklmnopqrstuvwxyz";
    private const string UppercaseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private const string DigitChars = "0123456789";
    private const string SpecialChars = "!@#$%^&*()-_=+[]{}|;:,.<>?";

    /// <summary>
    /// Generates a cryptographically secure random password that satisfies all policy requirements.
    /// </summary>
    /// <param name="length">Desired password length (minimum 12, default 16).</param>
    /// <returns>A random password meeting all complexity requirements.</returns>
    public static string GenerateSecurePassword(int length = 16)
    {
        if (length < 12)
            length = 12;

        using var rng = System.Security.Cryptography.RandomNumberGenerator.Create();
        var password = new char[length];
        var allChars = LowercaseChars + UppercaseChars + DigitChars + SpecialChars;

        password[0] = GetRandomChar(rng, LowercaseChars);
        password[1] = GetRandomChar(rng, UppercaseChars);
        password[2] = GetRandomChar(rng, DigitChars);
        password[3] = GetRandomChar(rng, SpecialChars);

        for (var i = 4; i < length; i++)
        {
            password[i] = GetRandomChar(rng, allChars);
        }

        Shuffle(rng, password);

        return new string(password);
    }

    private static char GetRandomChar(System.Security.Cryptography.RandomNumberGenerator rng, string chars)
    {
        var bytes = new byte[4];
        rng.GetBytes(bytes);
        var index = (int)(BitConverter.ToUInt32(bytes, 0) % (uint)chars.Length);
        return chars[index];
    }

    private static void Shuffle(System.Security.Cryptography.RandomNumberGenerator rng, char[] array)
    {
        var bytes = new byte[4];
        for (var i = array.Length - 1; i > 0; i--)
        {
            rng.GetBytes(bytes);
            var j = (int)(BitConverter.ToUInt32(bytes, 0) % (uint)(i + 1));
            (array[i], array[j]) = (array[j], array[i]);
        }
    }
}
